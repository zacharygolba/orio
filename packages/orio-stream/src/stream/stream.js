// @flow

import { AsAsyncIterator, ToString } from 'orio-traits'
import { identity, reduceAsync } from 'orio-utils'
import type {
  AsyncDest,
  AsyncProducer,
  AsyncIteratorResult,
  FromIterator,
  AsyncSource,
} from 'orio-types'

import * as sink from '../sink'
import type { Sink } from '../sink'

import {
  Chain,
  Concat,
  Cycle,
  Enumerate,
  Filter,
  FilterMap,
  FlatMap,
  Map,
  Skip,
  SkipWhile,
  Take,
  TakeWhile,
  Tap,
  Unique,
  Zip,
} from './adapter'

@ToString
@AsAsyncIterator
export default class Stream<T> implements AsyncProducer<T> {
  /*:: @@asyncIterator: () => $AsyncIterator<T, void, void> */
  producer: AsyncProducer<T>

  constructor(producer: AsyncProducer<T>) {
    this.producer = producer
  }

  chain<U>(source: AsyncSource<U>): Stream<T | U> {
    const adapter = new Chain(this.producer, source)
    return new Stream(adapter)
  }

  concat<U>(source: AsyncSource<U>): Stream<T | U> {
    const adapter = new Concat(this.producer, source)
    return new Stream(adapter)
  }

  async collect<C: FromIterator<T>>(Target?: Class<C>): Promise<C> {
    const items = await this.reduce((acc, next) => {
      acc.push(next)
      return acc
    }, [])

    if (Target === Array || Target == null) {
      return items
    }

    return new Target(items)
  }

  count(): Promise<number> {
    return this.reduce(acc => acc + 1, 0)
  }

  cycle(): Stream<T> {
    const adapter = new Cycle(this.producer)
    return new Stream(adapter)
  }

  drop(): void {
    this.producer.drop()
  }

  enumerate(): Stream<[number, T]> {
    const adapter = new Enumerate(this.producer)
    return new Stream(adapter)
  }

  async every(fn: T => Promise<boolean> | boolean): Promise<boolean> {
    const item = await this.find(async i => !await fn(i))
    return item === undefined
  }

  filterMap<U>(fn: T => Promise<?U> | ?U): Stream<U> {
    const adapter = new FilterMap(this.producer, fn)
    return new Stream(adapter)
  }

  filter(fn: T => Promise<boolean> | boolean): Stream<T> {
    const adapter = new Filter(this.producer, fn)
    return new Stream(adapter)
  }

  find(fn: T => Promise<boolean> | boolean): Promise<?T> {
    return this.filter(fn).first()
  }

  first(): Promise<?T> {
    return this.take(1).reduce((_, next) => next)
  }

  flatMap<U>(fn: T => AsyncSource<U>): Stream<U> {
    const adapter = new FlatMap(this.producer, fn)
    return new Stream(adapter)
  }

  flatten(): Stream<*> {
    return this.flatMap(identity)
  }

  async forEach(fn: T => Promise<void> | void): Promise<void> {
    await this.reduce((_, item) => fn(item))
  }

  join(sep?: string = ','): Promise<string> {
    return this.reduce((prev, next) => {
      if (prev.length > 0) {
        return prev + sep + String(next)
      }

      return String(next)
    }, '')
  }

  last(): Promise<?T> {
    return this.reduce((_, next) => next)
  }

  map<U>(fn: T => Promise<U> | U): Stream<U> {
    const adapter = new Map(this.producer, fn)
    return new Stream(adapter)
  }

  async next(): AsyncIteratorResult<T, void> {
    const next = await this.producer.next().catch(e => {
      this.drop()
      throw e
    })

    if (next.done) {
      this.drop()
    }

    return next
  }

  async nth(idx: number): Promise<?T> {
    const [lastIdx, value] = await this.enumerate()
      .take(idx + 1)
      .last()
      .then(item => item || [])

    if (idx === lastIdx) {
      return value
    }

    return undefined
  }

  pipe(dest: AsyncDest): Sink<*> {
    return sink.from(dest).push(this)
  }

  async product(): Promise<number> {
    const [lhs, rhs] = await Promise.all([
      this.next().then(({ value }) => +value || 0),
      this.next().then(({ value }) => +value || 0),
    ])

    return this.reduce((acc, item) => acc * +item, lhs * rhs)
  }

  reduce<U>(fn: (U, T) => Promise<U> | U, acc: Promise<U> | U): Promise<U> {
    return reduceAsync(fn, acc, this)
  }

  skip(amount: number): Stream<T> {
    const adapter = new Skip(this.producer, amount)
    return new Stream(adapter)
  }

  skipWhile(fn: T => Promise<boolean> | boolean): Stream<T> {
    const adapter = new SkipWhile(this.producer, fn)
    return new Stream(adapter)
  }

  async some(fn: T => Promise<boolean> | boolean): Promise<boolean> {
    const item = await this.find(fn)
    return item !== undefined
  }

  sum(): Promise<number> {
    return this.reduce((acc, item) => acc + +item, 0)
  }

  take(amount: number): Stream<T> {
    const adapter = new Take(this.producer, amount)
    return new Stream(adapter)
  }

  takeWhile(fn: T => Promise<boolean> | boolean): Stream<T> {
    const adapter = new TakeWhile(this.producer, fn)
    return new Stream(adapter)
  }

  tap(fn: T => Promise<void> | void): Stream<T> {
    const adapter = new Tap(this.producer, fn)
    return new Stream(adapter)
  }

  unique<U>(fn?: T => Promise<U> | Promise<T> | U | T = identity): Stream<T> {
    const adapter = new Unique(this.producer, fn)
    return new Stream(adapter)
  }

  zip(source?: AsyncSource<*> = this.producer): Stream<[T, *]> {
    const adapter = new Zip(this.producer, source)
    return new Stream(adapter)
  }
}
