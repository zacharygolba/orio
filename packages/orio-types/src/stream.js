// @flow

import type { Readable, Writable } from 'stream'

import type { Drop } from './drop'
import type { AsyncIterator, AsyncIteratorResult, Source } from './iterator'

export type AsyncDest = WritableStream | Writable
export type AsyncSource<T> =
  | Promise<Source<T>>
  | AsyncIterable<T>
  | ReadableStream
  | Readable
  | Source<T>

export interface AsyncConsumer<-T> extends Drop {
  write(chunk: T): Promise<void>,
}

export interface AsyncProducer<+T>
  extends AsyncIterable<T>, AsyncIterator<T>, Drop {}

export interface ReadableSource<+T> {
  cancel(reason?: string): void,
  read(): AsyncIteratorResult<T, void>,
}

export interface WritableDest<-T> {
  abort(reason?: string): void,
  write(chunk: T): Promise<void>,
}
