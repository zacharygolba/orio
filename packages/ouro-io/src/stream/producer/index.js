// @flow

import { intoAsyncIterator, intoIterator } from 'ouro-utils'
import type { AsyncProducer, AsyncSource } from 'ouro-types'

import * as nodeStream from '../../shims/node'
import * as webStream from '../../shims/web'
import Stream from '../stream'

import Chars from './chars'
import Empty from './empty'
import Indexed from './indexed'
import Once from './once'
import Reader from './reader'
import Task from './task'
import Unbound from './unbound'

export { Chars, Empty, Indexed, Once, Reader, Task, Unbound }

export function createProducer<T>(source: AsyncSource<T>): AsyncProducer<T> {
  if (source == null) {
    return new Empty()
  }

  if (source instanceof Promise) {
    return new Task(source)
  }

  if (source instanceof Stream) {
    return source.producer
  }

  if (webStream.isReadable(source)) {
    return new Reader(webStream.getReader(source))
  }

  if (nodeStream.isReadable(source)) {
    return new Reader(nodeStream.wrapReadable(source))
  }

  {
    const iterator = intoAsyncIterator(source)

    if (iterator != null) {
      return new Unbound(iterator)
    }
  }

  if (Array.isArray(source)) {
    return new Indexed(source)
  }

  if (typeof source === 'string') {
    return new Chars(source)
  }

  {
    const iterator = intoIterator(source)

    if (iterator != null) {
      return new Unbound(iterator)
    }
  }

  return new Once((source: any))
}
