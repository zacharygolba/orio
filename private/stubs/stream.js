// @flow

export class Duplex {}
export class PassThrough {}
export class Readable {}
export class Transform {}
export class Writable {}
