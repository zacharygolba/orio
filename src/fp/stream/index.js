// @flow

import curry from '../../utils/curry'

import * as adapters from './adapters'
import * as utils from './utils'

export type { Filter, Map } from './adapters'
export { enumerate } from './adapters'
export const chain = curry(adapters.chain)
export const filterMap = curry(adapters.filterMap)
export const filter = curry(adapters.filter)
export const flatMap = curry(adapters.flatMap)
export const map = curry(adapters.map)
export const skipWhile = curry(adapters.skipWhile)
export const skip = curry(adapters.skip)
export const takeWhile = curry(adapters.takeWhile)
export const take = curry(adapters.take)
export const tap = curry(adapters.tap)
export const zip = curry(adapters.zip)

export * from './from'
export { default as range } from './range'
export { default as repeat } from './repeat'

export type { ForEach, Reduce } from './utils'
export { count, first, last, reduce, sum, product } from './utils'
export const every = curry(utils.every)
export const find = curry(utils.find)
export const forEach = curry(utils.forEach)
export const join = curry(utils.join)
export const nth = curry(utils.nth)
export const some = curry(utils.some)
