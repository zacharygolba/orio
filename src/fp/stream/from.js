// @flow

import { Readable } from 'stream'
import type { EventEmitter } from 'events'

import noop from '../../utils/noop'

import { map } from './adapters'

const once = (source, event) => new Promise(resolve => {
  source.once(event, resolve)
})

export function from<T>(source: T): AsyncIterator<T> {
  return map(noop, source)
}

export async function* fromEvent(source: EventEmitter, event: string): AsyncIterator<*> {
  for (;;) {
    // eslint-disable-next-line no-await-in-loop
    yield await once(source, event)
  }
}

export async function* fromReadable(source: Readable): AsyncIterator<Buffer | string> {
  let done = false
  const end = () => { done = true }

  source.once('end', end)
  source.once('error', end)
  source.once('close', end)

  while (!done) {
    // eslint-disable-next-line no-await-in-loop
    yield await once(source, 'data')
    console.log(done)
  }
}
