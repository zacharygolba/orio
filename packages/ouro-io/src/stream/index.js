// @flow

import type { AsyncSource } from 'ouro-types'

import Stream from './stream'
import { createProducer, Empty } from './producer'

export type { Stream }

export function empty(): Stream<any> {
  const producer = new Empty()
  return new Stream(producer)
}

export function from<T>(source?: AsyncSource<T>): Stream<T> {
  const producer = createProducer(source)
  return new Stream(producer)
}
