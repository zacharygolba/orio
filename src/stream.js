import * as stream from './fp/stream'

class Stream {
  constructor(source) {
    this.source = source
  }

  [Symbol.iterator]() {
    return this
  }

  chain(target) {
    this.source = stream.chain(target)(this.source)
    return this
  }

  count() {
    return stream.count(this)
  }

  enumerate() {
    this.source = stream.enumerate(this.source)
    return this
  }

  every(fn) {
    return stream.every(fn)(this)
  }

  filter(fn) {
    this.source = stream.filter(fn)(this.source)
    return this
  }

  filterMap(fn) {
    this.source = stream.filterMap(fn)(this.source)
    return this
  }

  find(fn) {
    return stream.find(fn)(this)
  }

  first() {
    return stream.first(this)
  }

  flatMap(fn) {
    this.source = stream.flatMap(fn)(this.source)
    return this
  }

  forEach(fn) {
    return stream.forEach(fn)(this)
  }

  join(sep) {
    return stream.join(sep)(this)
  }

  last() {
    return stream.last(this)
  }

  map(fn) {
    this.source = stream.map(fn)(this.source)
    return this
  }

  next() {
    return this.source.next()
  }

  nth(index) {
    return stream.nth(index)(this)
  }

  product() {
    return stream.product(this)
  }

  reduce(fn, init) {
    return stream.reduce(fn, init, this)
  }

  skip(amount) {
    this.source = stream.skip(amount)(this.source)
    return this
  }

  skipWhile(fn) {
    this.source = stream.skipWhile(fn)(this.source)
    return this
  }

  some(fn) {
    return stream.some(fn)(this)
  }

  sum() {
    return stream.sum(this)
  }

  take(amount) {
    this.source = stream.take(amount)(this.source)
    return this
  }

  takeWhile(fn) {
    this.source = stream.takeWhile(fn)(this.source)
    return this
  }

  tap(fn) {
    this.source = stream.tap(fn)(this.source)
    return this
  }

  zip(target) {
    this.source = stream.zip(target)(this.source)
    return this
  }
}

export function from(source) {
  return new Stream(stream.from(source))
}

export function fromEvent(source, event) {
  return new Stream(stream.fromEvent(source, event))
}

export function fromReadable(source) {
  return new Stream(stream.fromReadable(source))
}

export function range(start, end) {
  return new Stream(stream.range(start, end))
}

export function repeat(value) {
  return new Stream(stream.repeat(value))
}
