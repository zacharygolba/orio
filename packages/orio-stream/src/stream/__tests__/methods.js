// @flow

import { Writable } from 'stream'

import * as stream from '../'
import Sink from '../../sink/sink'

test('#collect()', async () => {
  expect(await stream.from('test').collect()).toMatchSnapshot()
  expect(await stream.from('test').collect(Set)).toMatchSnapshot()
  expect(
    await stream
      .from('test')
      .enumerate()
      .collect(Map),
  ).toMatchSnapshot()
})

test('#count()', async () => {
  expect(await stream.from([1, 2, 3, 4]).count()).toBe(4)
})

test('#drop()', () => {
  expect(stream.empty().drop()).toBeUndefined()
})

test('#every()', async () => {
  {
    const fn = jest.fn(num => num < 10)

    expect(await stream.from([1, 2, 3, 4, 5]).every(fn)).toBe(true)
    expect(fn).toHaveBeenCalledTimes(5)
  }

  {
    const fn = jest.fn(num => num > 10)

    expect(await stream.from([1, 2, 3, 4, 5]).every(fn)).toBe(false)
    expect(fn).toHaveBeenCalledTimes(1)
  }
})

test('#find()', async () => {
  const fn = jest.fn(char => char === 'c')

  expect(await stream.from(['a', 'b', 'c', 'd']).find(fn)).toBe('c')
  expect(fn).toHaveBeenCalledTimes(3)
})

test('#first()', async () => {
  expect(await stream.from(['a', 'b', 'c']).first()).toBe('a')
})

test('#forEach()', async () => {
  {
    const source = [1, 2, 3]
    const subj = stream.from(source)
    const fn = jest.fn()

    await subj.forEach(fn)
    source.forEach(value => expect(fn).toHaveBeenCalledWith(value))
    expect(fn).toHaveBeenCalledTimes(source.length)
  }

  {
    const subj = stream.from([1, 2, 3, 4])
    const fn = jest.fn()

    await subj.filter(num => num % 2 === 0).forEach(fn)
    expect(fn).toHaveBeenCalledWith(2)
    expect(fn).toHaveBeenCalledWith(4)
    expect(fn).toHaveBeenCalledTimes(2)
  }
})

test('#join()', async () => {
  expect(await stream.from([1, 2, 3]).join()).toBe('1,2,3')
  expect(await stream.from([1, 2, 3]).join(' * ')).toBe('1 * 2 * 3')
})

test('#last()', async () => {
  expect(await stream.from(['a', 'b', 'c']).last()).toBe('c')
})

describe('#next()', () => {
  let subj

  beforeEach(() => {
    subj = stream.from('test')
    jest.spyOn(subj.producer, 'drop')
  })

  test('success', async () => {
    expect(await subj.next()).toMatchSnapshot()
    expect(await subj.next()).toMatchSnapshot()
    expect(await subj.next()).toMatchSnapshot()
    expect(await subj.next()).toMatchSnapshot()
    expect(await subj.next()).toMatchSnapshot()
    expect(subj.producer.drop).toHaveBeenCalledTimes(1)
  })

  test('failure', async () => {
    await subj
      .map(() => {
        throw new Error('error')
      })
      .next()
      .catch(e => {
        expect(e).toBeInstanceOf(Error)
        expect(subj.producer.drop).toHaveBeenCalled()
      })
  })
})

test('#nth()', async () => {
  expect(await stream.from(['a', 'b', 'c']).nth(0)).toBe('a')
  expect(await stream.from(['a', 'b', 'c']).nth(1)).toBe('b')
  expect(await stream.from(['a', 'b', 'c']).nth(2)).toBe('c')
  expect(await stream.from(['a', 'b', 'c']).nth(-1)).toBeUndefined()
  expect(await stream.from(['a', 'b', 'c']).nth(4)).toBeUndefined()
})

test('#pipe()', () => {
  const dest = new Writable({
    write: jest.fn(),
    writev: jest.fn(),
  })

  expect(stream.empty().pipe(dest)).toBeInstanceOf(Sink)
})

test('#product()', async () => {
  expect(await stream.empty().product()).toBe(0)
  expect(await stream.from([2, 3, 4]).product()).toBe(2 * 3 * 4)
  expect(Number.isNaN(await stream.from(['a', 'b', 'c']).product())).toBe(true)
})

test('#reduce()', async () => {
  const set = await stream
    .from([1, 2, 3])
    .reduce((acc, next) => acc.add(next), new Set())

  expect(set.size).toBe(3)
  expect(set.has(1)).toBe(true)
  expect(set.has(2)).toBe(true)
  expect(set.has(3)).toBe(true)
})

test('#some()', async () => {
  {
    const fn = jest.fn(num => num > 2)

    expect(await stream.from([1, 2, 3, 4, 5]).some(fn)).toBe(true)
    expect(fn).toHaveBeenCalledTimes(3)
  }

  {
    const fn = jest.fn(num => num > 10)

    expect(await stream.from([1, 2, 3, 4, 5]).some(fn)).toBe(false)
    expect(fn).toHaveBeenCalledTimes(5)
  }
})

test('#sum()', async () => {
  expect(await stream.empty().sum()).toBe(0)
  expect(await stream.from([1, 2, 3, 4]).sum()).toBe(1 + 2 + 3 + 4)
  expect(Number.isNaN(await stream.from(['a', 'b', 'c']).sum())).toBe(true)
})

test('#toString()', () => {
  expect(stream.empty().toString()).toBe('[object Stream]')
})
