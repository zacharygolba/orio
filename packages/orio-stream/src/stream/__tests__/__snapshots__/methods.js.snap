// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#collect() 1`] = `
Array [
  "t",
  "e",
  "s",
  "t",
]
`;

exports[`#collect() 2`] = `
Set {
  "t",
  "e",
  "s",
}
`;

exports[`#collect() 3`] = `
Map {
  0 => "t",
  1 => "e",
  2 => "s",
  3 => "t",
}
`;

exports[`#next() success 1`] = `
Object {
  "done": false,
  "value": "t",
}
`;

exports[`#next() success 2`] = `
Object {
  "done": false,
  "value": "e",
}
`;

exports[`#next() success 3`] = `
Object {
  "done": false,
  "value": "s",
}
`;

exports[`#next() success 4`] = `
Object {
  "done": false,
  "value": "t",
}
`;

exports[`#next() success 5`] = `
Object {
  "done": true,
  "value": undefined,
}
`;
